### PR TITLE
[Merged by Bors] - feat(Algebra/Lie/Weights/Chain): More API for chains.

### PR DESCRIPTION
**Pull request by erdOne**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Lie/Weights/Chain.lean [hunk 1]
```diff
@@ -217,4 +217,124 @@ end IsCartanSubalgebra
...
+lemma weightSpace_neg_zsmul_add_ne_bot {n : ℕ} (hn : n ≤ chainBotCoeff α β) :
+    weightSpace M ((-n : ℤ) • α + β : L → R) ≠ ⊥ := by
+  apply weightSpace_zsmul_add_ne_bot α β <;> omega
+
+/-- This is the first weight in an `α`-chain through `β`. -/
+noncomputable
+def chainTop (α : L → R) (β : Weight R L M) : Weight R L M :=
+  ⟨chainTopCoeff α β • α + β, weightSpace_nsmul_add_ne_bot_of_le α β le_rfl⟩
+
+/-- This is the first weight in an `α`-chain through `β`. -/
```
Comment (jcommelin) [inline Mathlib/Algebra/Lie/Weights/Chain.lean:303]: These two docstrings are identical. I guess you want `s/first/last/` in the first docstring?

Comment (jcommelin) [reply Mathlib/Algebra/Lie/Weights/Chain.lean:303]: Also, I would start the docstrings with `The .. weight`.
```suggestion
/-- The last weight in an `α`-chain through `β`. -/
noncomputable
def chainTop (α : L → R) (β : Weight R L M) : Weight R L M :=
  ⟨chainTopCoeff α β • α + β, weightSpace_nsmul_add_ne_bot_of_le α β le_rfl⟩

/-- The first weight in an `α`-chain through `β`. -/
```

File: Mathlib/Algebra/Lie/Weights/Chain.lean [hunk 2]
```diff
@@ -217,4 +217,124 @@ end IsCartanSubalgebra
...
+
+lemma chainTop_isNonZero' (hα' : weightSpace M α ≠ ⊥) :
+    (chainTop α β).IsNonZero := by
+  by_contra e
+  apply hα'
+  rw [← add_zero (α : L → R), ← e, weightSpace_add_chainTop _ _ hα]
+
+lemma chainTop_isNonZero (α β : Weight R L M) (hα : α.IsNonZero) :
+    (chainTop α β).IsNonZero :=
+  chainTop_isNonZero' α β hα α.2
```
Comment (jcommelin) [inline Mathlib/Algebra/Lie/Weights/Chain.lean:336]: Should we have analogous lemmas for `chainBot`?

Comment (erdOne) [reply Mathlib/Algebra/Lie/Weights/Chain.lean:336]: I don't think this is possible in the general case. For example `chainBot α α` will be zero if `-α` is not a weight (or will this never happen)?

File: Mathlib/Algebra/Lie/Weights/Chain.lean [hunk 3]
```diff
@@ -217,4 +217,124 @@ end IsCartanSubalgebra
...
+@[simp] lemma chainBot_neg : chainBot (-α) β = chainTop α β := by ext; simp
+
+@[simp] lemma chainTop_zero : chainTop 0 β = β := by ext; simp
+@[simp] lemma chainBot_zero : chainBot 0 β = β := by ext; simp
+
+lemma weightSpace_add_chainTop :
+    weightSpace M (α + chainTop α β : L → R) = ⊥ := by
+  rw [coe_chainTop', ← add_assoc, ← succ_nsmul', weightSpace_chainTopCoeff_add_one_nsmul_add _ _ hα]
+
+lemma weightSpace_neg_add_chainTop :
```
Comment (jcommelin) [inline Mathlib/Algebra/Lie/Weights/Chain.lean:324]:
```suggestion
lemma weightSpace_neg_add_chainBot :
```

File: Mathlib/Algebra/Lie/Weights/Chain.lean (hunk 2 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/Weights/Chain.lean]: Fair enough

bors d+

**Comment by erdOne** [comment]
bors merge

**Pull request closed without merging**